File: frontend/src/types/invoice.ts
export interface Invoice {
  id: string;
  workgroup_id: string;
  contractor_id: string;
  invoice_number: string; // sequential per workgroup (INV-001, INV-002)
  amount: number;
  line_items: InvoiceLineItem[];
  status: InvoiceStatus;
  ai_validated: boolean;
  ai_flags: AIFlag[];
  approved_by?: string;
  approved_at?: string;
  file_url?: string;
  notes?: string;
  submitted_at: string;
  paid_at?: string;
  created_at: string;
  updated_at: string;
}

export interface InvoiceLineItem {
  job_id: string;
  job_title: string;
  amount: number;
  job_budget: number;
  variance: number; // amount - job_budget
  description?: string;
}

export type InvoiceStatus =
  | 'draft'
  | 'submitted'
  | 'ai_validated'
  | 'ai_flagged'
  | 'pending_approval'
  | 'approved'
  | 'rejected'
  | 'paid';

export interface AIFlag {
  type: 'double_billing' | 'budget_exceeded' | 'job_incomplete' | 'duplicate' | 'mismatch' | 'other';
  message: string;
  severity: 'warning' | 'error';
  job_id?: string;
}

export interface SubmitInvoiceRequest {
  line_items: {
    job_id: string;
    amount: number;
    description?: string;
  }[];
  notes?: string;
  file_url?: string;
}

export interface WorkgroupInvoiceSummary {
  workgroup_id: string;
  workgroup_budget: number;
  total_invoiced: number;
  total_paid: number;
  remaining: number;
  invoices: Invoice[];
  uninvoiced_jobs: {
    job_id: string;
    job_title: string;
    budget: number;
    status: string;
  }[];
}

File: backend/app/models/invoice.py
"""
Pydantic Models — Invoice

Invoices belong to a Workgroup, contain line items referencing Jobs.
Multiple invoices per workgroup allowed (invoice as you go).
Each job can appear on only ONE invoice (no double-billing).
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from decimal import Decimal
from enum import Enum


class InvoiceStatus(str, Enum):
    DRAFT = "draft"
    SUBMITTED = "submitted"
    AI_VALIDATED = "ai_validated"
    AI_FLAGGED = "ai_flagged"
    PENDING_APPROVAL = "pending_approval"
    APPROVED = "approved"
    REJECTED = "rejected"
    PAID = "paid"


# ── Line Items ────────────────────────────────────────

class InvoiceLineItem(BaseModel):
    """A single line item on an invoice, referencing a Job."""
    job_id: str
    job_title: str
    amount: Decimal = Field(..., ge=0, decimal_places=2)
    description: Optional[str] = None


# ── Request Models ────────────────────────────────────

class InvoiceCreate(BaseModel):
    workgroup_id: str
    invoice_number: str
    line_items: list[InvoiceLineItem]
    file_url: Optional[str] = None      # optional PDF upload

class InvoiceApproval(BaseModel):
    action: str = Field(..., pattern="^(approve|reject|query)$")
    reason: Optional[str] = None        # required for reject/query


# ── Response Models ───────────────────────────────────

class InvoiceResponse(BaseModel):
    id: str
    workgroup_id: str
    contractor_id: str
    submitted_by_worker: Optional[str] = None
    invoice_number: str
    amount: Decimal
    line_items: list[InvoiceLineItem] = []
    status: InvoiceStatus
    ai_validated: bool = False
    ai_flags: list = []
    approved_by: Optional[str] = None
    approved_at: Optional[datetime] = None
    file_url: Optional[str] = None
    submitted_at: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

class InvoiceValidationResult(BaseModel):
    """Result of AI invoice validation (10-point checklist)."""
    is_valid: bool
    checks: list[dict] = []             # { check: str, passed: bool, detail: str }
    flags: list[str] = []
    recommendation: str = ""            # "route_to_approval" | "flag_for_review"
